frontend/src/diagram/sprotty/utils/geometry.tsx

```diff
@@ -0,0 +1,111 @@
+/*******************************************************************************
+ * Copyright (c) 2022 THALES GLOBAL SERVICES.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import { Bounds, Point } from 'sprotty';
+
+export enum RectangleSide {
+  'north',
+  'south',
+  'west',
+  'east',
+}
+/**
+ * Returns the distance between the point p and the segment made of p1 and p2
+ * @param p the point
+ * @param p1 the first point of the segment
+ * @param p2 the second point of the segment
+ */
+export function closestPointOnSegment(p: Point, p1: Point, p2: Point): Point {
+  var A = p.x - p1.x;
+  var B = p.y - p1.y;
+  var C = p2.x - p1.x;
+  var D = p2.y - p1.y;
+
+  var dot = A * C + B * D;
+  var lengthSquare = C * C + D * D;
+  // param is the projection distance from p1 on the segment. This is the fraction of the segment that p is closest to.
+  var param = -1;
+  if (lengthSquare != 0)
+    //if the line has 0 length
+    param = dot / lengthSquare;
+
+  var nearestPoint: Point = p1;
+
+  if (param < 0) {
+    // The point is not "in front of" the segment. It is "away" on the p1 side
+    nearestPoint = { x: p1.x, y: p1.y };
+  } else if (param > 1) {
+    // The point is not "in front of" the segment. It is "away" on the p1 side
+    nearestPoint = { x: p2.x, y: p2.y };
+  } else {
+    // The point is "in front of" the segment
+    // get the projection of p on the segment
+    nearestPoint = { x: p1.x + param * C, y: p1.y + param * D };
+  }
+
+  return nearestPoint;
+}
+
+export interface SnapToRectangleInfo {
+  pointOnRectangleSide: Point;
+  side: RectangleSide;
+}
+
+/**
+ * Returns a point on the rectangle that is the closest from the given p point.
+ */
+export const snapToRectangle = (point: Point, rectangle: Bounds): SnapToRectangleInfo => {
```

Comment:
This code seems nice but could you:

- Add an ADR describing very quickly the strategy that you have selected for your algorithm
- Mention said ADR in the CHANGELOG
- Add an entry in the CHANGELOG for this contribution
- Add some unit tests (see the tests of `convertDiagram.ts`) because this seems very easy to test

Reply:
There is still the test to do but you can already review the rest of your remarks

Reply:
I added a previous commit fix the diagram convertion fromGQL to sprotty so that I could do the proper test that I also added